prime-router/docs/design/design/auth/auth-design.md

```diff
@@ -0,0 +1,185 @@
+
+## Reasons for updating
+
+- Unify authn/authz to one consistent method as having multiple methods makes it difficult to determine who is using which method
+- Reduce attack surface area since currently we have the weaknesses of each method
+- Minimize authn/authz issues being the cause for missing sent reports
+- Make life easier for Engagement team by having a single choice when onboarding a new sender/receiver
+
+## Requirements
+
+- Protect certain endpoints from only being accessed by those we trust
+- Centralize user management
+- Revoke access if a secret is compromised
+- Audit which clients are accessing which resources
+- Fine-tune access control for specific clients
+- Unify authn/authz approach for all microservices
+- Allow clients the ability to self-service to ease client onboarding
+
+## Identity Provider Options
+
+### Azure AD
+
+Pros
+- Integrated into the Microsoft ecosystem that we currently reside
+- Scales well
+
+Cons
+- Complex setup
+ 
+### Okta
+
+Pros
+- A trusted industry standard
+- Highly scalable
+- Offers self-service to ease Engagement teams workload
+- We already have many of our clients set up here!
+
+Cons
+- Single point of failure if there is an outage
+
+### Homegrown
+
+Pros
+- As flexible as we need it to be
+- Cheap
+
+Cons
+- Custom authn/authz code is often vulnerable to attacks we may not have considered
+- Additional work for us to write and maintain
+
+## Protocol
+
+We will be using OAuth 2.0 with OIDC. OAuth 2.0 is an industry-standard authorization protocol and OIDC is
+another open standard on top of OAuth that adds an identity layer (authentication).
+
+Azure and Okta both support this protocol but if we were to go with the homegrown I would still highly recommend 
+following the spec.
+
+## Architecture options
+
+### Spring reverse proxy microservice
+
+This approach would create a new microservice whose sole responsibility is authn/authz. It would sit in front of
+all requests to protected endpoints.
+
+### Prerequisites
+- Designate protected endpoints and put them behind a VPC.
+  - Traffic must flow through our authn/authz service  
+  - Protected endpoints should not be accessible to the public internet
+
+A typical request flow would look like this
+- Incoming request to protected endpoint hits our authn/authz service
+- Depending on our implementation, we decide if that request is legitimate and allowed to access the designated endpoint
+  - In cloud-based solutions, we would make calls out to an external IdP
+  - In a homegrown solution, we would introspect certain cookies and headers and compare them to what we have in a datastore
+- If the request is unauthenticated, forward the user along to a login page (often provided by the external IdP)
+- If the request is unauthorized, short-circuit the request to a 401 unauthorized
+- If the request is authenticated and authorized for the resource, pass it along to the internal protected endpoint
+
+Pros
```

Comment:
Can this reverse proxy be configured for other endpoints outside of core RS microservices, like ETOR endpoints? If so, let's add that as a pro because it would allow RS users and machines to access ETOR endpoints without having to auth separately

Reply:
It can be configured for anything we want!